app.go

```diff
@@ -2,29 +2,81 @@ package passage
 
 import (
 	"crypto/rsa"
+	"encoding/json"
 	"errors"
+	"io/ioutil"
 	"net/http"
 
 	"github.com/dgrijalva/jwt-go"
 )
 
 type App struct {
-	Handle    string
-	PublicKey *rsa.PublicKey
+	handle    string
+	publicKey *rsa.PublicKey
 }
 
-func New() *App {
-	return &App{}
-}
+var apiKey string
+var publicKeyCache map[string]*rsa.PublicKey = make(map[string]*rsa.PublicKey)
 
-func (a *App) AuthenticateRequest(r *http.Request) (*User, error) {
-	// Check if the app's public key is set. If not, attempt to set it.
-	if a.PublicKey == nil {
-		pk, err := getPublicKeyFromEnv()
+func New(appHandle string, key ...string) (*App, error) {
+	if len(key) > 0 {
+		apiKey = key[0]
+	}
+	var pubKey *rsa.PublicKey
+	if _, ok := publicKeyCache[appHandle]; ok {
+		pubKey = publicKeyCache[appHandle]
+	} else {
+		var pubKey *rsa.PublicKey
+		//MAKE SOME REQUEST TO PASSAGE
+		resp, err := getAppInfo(appHandle)
 		if err != nil {
 			return nil, err
 		}
-		a.PublicKey = pk
+		pubKey, err = getRSAPublicKey(resp.PublicKey)
+		if err != nil {
+			return nil, err
+		}
+		publicKeyCache[appHandle] = pubKey
+	}
+	return &App{
+		handle:    appHandle,
+		publicKey: pubKey,
+	}, nil
+}
+
+type appInfoResp struct {
+	PublicKey string `json:"public_key"`
+}
+
+func getAppInfo(appHandle string) (*appInfoResp, error) {
```

Comment:
Since this is only handling public keys maybe rename to `getAppPublicKey` ?

and just return a base-64 encoded public key string?

I like the flexibility of the appInfo and appInfoResp, but maybe we keep it simple until we need more complexity?